src/brod_consumer.erl

```diff
@@ -292,123 +235,130 @@ code_change(_OldVsn, State, _Extra) ->
-      NewOffset0 = State#state.offset,
-      SocketPid = State#state.socket_pid,
-      {ok, NewOffset} = maybe_fetch_valid_offset(SocketPid, OldOffset,
-                                                 NewOffset0, Topic, Partition),
-      maybe_delay_fetch_request(NextFetchTime),
-      {noreply, State#state{ cb_state   = NewCbState
-                           , offset     = NewOffset
-                           , last_error = undefined
-                           }};
+handle_fetch_response(_Response, _CorrId,
+                      #state{subscriber = undefined} = State) ->
+  %% discard fetch response when there is no (dead?) subscriber
+  {noreply, State};
+handle_fetch_response(_Response, CorrId1,
+                      #state{ last_corr_id = CorrId2
+                            } = State) when CorrId1 < CorrId2 ->
+  %% handle obsolete fetch responses in case we got new offset from callback
+  error_logger:info_msg("~p ~p Dropping obsolete fetch response "
+                        "with corr_id = ~p",
+                        [?MODULE, self(), CorrId1]),
+  {noreply, State};
+handle_fetch_response(#fetch_response{ topics = []
+                                     , error = max_bytes_too_small
+                                     }, _CorrId, State) ->
+  %% TODO try with a larger max_bytes
+  {noreply, State};
+handle_fetch_response(#fetch_response{ topics = [TopicFetchData]
+                                     , error  = undefined
+                                     }, CorrId, State) ->
+  CorrId = State#state.last_corr_id, %% assert
+  #topic_fetch_data{ topic = Topic
+                   , partitions = [PM]} = TopicFetchData,
+  #partition_messages{ partition = Partition
+                     , error_code = ErrorCode
+                     , high_wm_offset = HighWmOffset
+                     , messages = Messages} = PM,
+  case brod_kafka:is_error(ErrorCode) of
+    true ->
+      Error = #kafka_fetch_error{ topic      = Topic
+                                , partition  = Partition
+                                , error_code = ErrorCode
+                                , error_desc = brod_kafka_errors:desc(ErrorCode)
+                                },
+      handle_fetch_error(Error, State);
+    false ->
+      MsgSet = #kafka_message_set{ topic          = Topic
+                                 , partition      = Partition
+                                 , high_wm_offset = HighWmOffset
+                                 , messages       = Messages
+                                 },
+      handle_message_set(MsgSet, State)
+  end.
+
+handle_message_set(#kafka_message_set{messages = []}, State) ->
+  ok = maybe_delay_fetch_request(State),
+  {noreply, State};
+handle_message_set(#kafka_message_set{messages = Messages} = MsgSet,
+                   #state{ subscriber    = Subscriber
+                         , pending_acks  = PendingAcks
+                         } = State0) ->
+  ok = cast(Subscriber, MsgSet),
```

Comment:
I think we can include Ref and self() in this message, so that users can 'ack' using info from the message itself.